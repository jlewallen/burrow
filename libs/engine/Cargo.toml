[package]
name = "engine"
version = "0.1.0"
authors = ["Jacob Lewallen <jlewallen@gmail.com>"]
edition = "2021"

[lib]
doctest = false

[dependencies.kernel]
path = "../kernel"

[dependencies]
anyhow = "1.0.72"
chrono = { version = "0.4.26", features = ["serde"] }
serde = { version = "1.0.180", features = ["derive"] }
serde_json = "1.0.104"
tracing = "0.1.37"
tracing-subscriber = { version = "0.3.17", features = ["json"] }
