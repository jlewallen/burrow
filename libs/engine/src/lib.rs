mod identifiers;
mod internal;
mod perform;
mod users;

pub mod domain;
pub mod notifications;
pub mod sequences;
pub mod session;
pub mod storage;

pub use domain::*;
pub use notifications::*;
pub use session::*;

pub use users::model::{add_username_to_key, username_to_key};
