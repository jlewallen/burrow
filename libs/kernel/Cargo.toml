[package]
name = "kernel"
version = "0.1.0"
authors = ["Jacob Lewallen <jlewallen@gmail.com>"]
edition = "2021"

[lib]
doctest = false

[dependencies.replies]
path = "../replies"

[dependencies]
anyhow = "1.0.71"
thiserror = "1.0.40"
tracing = "0.1.37"
nom = "7.1.3"
serde_json = "1.0.96"
serde = { version = "1.0.163", features = ["derive"] }
treediff = { version = "4.0.2", features = ["with-serde-json"] }