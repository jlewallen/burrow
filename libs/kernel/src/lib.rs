pub mod actions;
pub mod hooks;
pub mod model;
pub mod perms;
pub mod plugins;
pub mod session;
pub mod surround;

pub mod common {
    pub use replies::*;
}

pub mod prelude {
    pub use replies::DomainEvent;

    pub use crate::actions::*;
    pub use crate::finder::*;
    pub use crate::hooks::*;
    pub use crate::model::compare::*;
    pub use crate::model::*;
    pub use crate::plugins::*;
    pub use crate::session::*;
    pub use crate::surround::*;
}

mod finder {
    use crate::model::{Audience, EntityKey, Entry, Item};
    use crate::surround::Surroundings;

    pub trait Finder: Send + Sync {
        fn find_world(&self) -> anyhow::Result<Entry>;

        fn find_location(&self, entry: &Entry) -> anyhow::Result<Entry>;

        fn find_item(
            &self,
            surroundings: &Surroundings,
            item: &Item,
        ) -> anyhow::Result<Option<Entry>>;

        fn find_audience(&self, audience: &Audience) -> anyhow::Result<Vec<EntityKey>>;
    }
}

#[macro_export]
macro_rules! here {
    () => {{
        format!("{}:{}", file!(), line!())
    }};
}

#[cfg(test)]
mod tests;
