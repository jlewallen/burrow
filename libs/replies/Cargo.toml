[package]
name = "replies"
version = "0.1.0"
authors = ["Jacob Lewallen <jlewallen@gmail.com>"]
edition = "2021"

[lib]
doctest = false

[dependencies.macros]
path = "../../libs/macros"

[dependencies.burrow-bon]
path = "../../libs/bon"

[dependencies]
anyhow = "1.0.72"
serde = { version = "1.0.180", features = ["derive"] }
serde_json = "1.0.104"
thiserror = "1.0.44"

[dev-dependencies]
insta = { version = "1.31.0", features = ["json"] }
