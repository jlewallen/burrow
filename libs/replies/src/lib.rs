use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::fmt::Debug;

pub trait ToJson: Debug {
    fn to_json(&self) -> Result<Value, serde_json::Error>;
}

pub trait Reply: ToJson {}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SimpleReply {
    Done,
    NotFound,
    What,
    Impossible,
    Prevented,
}

impl ToJson for SimpleReply {
    fn to_json(&self) -> Result<Value, serde_json::Error> {
        serde_json::to_value(BasicReply::Simple(self.clone()))
    }
}

impl Reply for SimpleReply {}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ObservedEntity {
    pub key: String,
    pub name: Option<String>,
    pub qualified: Option<String>,
    pub desc: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AreaObservation {
    pub area: ObservedEntity,
    pub person: ObservedEntity,
    pub living: Vec<ObservedEntity>,
    pub items: Vec<ObservedEntity>,
    pub carrying: Vec<ObservedEntity>,
    pub routes: Vec<ObservedEntity>,
}

impl Reply for AreaObservation {}

impl ToJson for AreaObservation {
    fn to_json(&self) -> Result<Value, serde_json::Error> {
        serde_json::to_value(BasicReply::AreaObservation(self.clone()))
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InsideObservation {
    pub vessel: ObservedEntity,
    pub items: Vec<ObservedEntity>,
}

impl Reply for InsideObservation {}

impl ToJson for InsideObservation {
    fn to_json(&self) -> Result<Value, serde_json::Error> {
        serde_json::to_value(BasicReply::InsideObservation(self.clone()))
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntityObservation {
    pub entity: ObservedEntity,
}

impl Reply for EntityObservation {}

impl ToJson for EntityObservation {
    fn to_json(&self) -> Result<Value, serde_json::Error> {
        serde_json::to_value(BasicReply::EntityObservation(self.clone()))
    }
}

pub trait Observed: ToJson {}

impl Observed for InsideObservation {}

impl Observed for AreaObservation {}

impl Observed for SimpleReply {}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SimpleObservation(serde_json::Value);

impl SimpleObservation {
    pub fn new(value: serde_json::Value) -> Self {
        Self(value)
    }
}

impl From<&SimpleObservation> for serde_json::Value {
    fn from(o: &SimpleObservation) -> Self {
        o.0.clone()
    }
}

impl ToJson for SimpleObservation {
    fn to_json(&self) -> Result<Value, serde_json::Error> {
        serde_json::to_value(BasicReply::SimpleObservation(self.clone()))
    }
}

impl Observed for SimpleObservation {}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum WorkingCopy {
    Description(String),
    Json(serde_json::Value),
    Script(String),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EditorReply {
    pub key: String,
    pub editing: WorkingCopy,
}

impl EditorReply {
    pub fn new(key: String, editing: WorkingCopy) -> Self {
        Self { key, editing }
    }
}

impl Reply for EditorReply {}

impl ToJson for EditorReply {
    fn to_json(&self) -> Result<Value, serde_json::Error> {
        serde_json::to_value(BasicReply::Editor(self.clone()))
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct JsonReply {
    value: serde_json::Value,
}

impl From<serde_json::Value> for JsonReply {
    fn from(value: serde_json::Value) -> Self {
        Self { value }
    }
}

impl Reply for JsonReply {}

impl ToJson for JsonReply {
    fn to_json(&self) -> Result<Value, serde_json::Error> {
        Ok(json!({ "json": self.value }))
    }
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum BasicReply {
    Simple(SimpleReply),
    EntityObservation(EntityObservation),
    InsideObservation(InsideObservation),
    AreaObservation(AreaObservation),
    SimpleObservation(SimpleObservation),
    Editor(EditorReply),
    Json(JsonReply),
}

impl ToJson for BasicReply {
    fn to_json(&self) -> Result<Value, serde_json::Error> {
        serde_json::to_value(self)
    }
}

impl From<SimpleReply> for BasicReply {
    fn from(value: SimpleReply) -> Self {
        Self::Simple(value)
    }
}

impl From<JsonReply> for BasicReply {
    fn from(value: JsonReply) -> Self {
        Self::Json(value)
    }
}

impl From<EditorReply> for BasicReply {
    fn from(value: EditorReply) -> Self {
        Self::Editor(value)
    }
}
