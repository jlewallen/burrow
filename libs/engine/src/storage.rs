use anyhow::Result;
use std::{
    collections::HashMap,
    rc::Rc,
    sync::{Arc, RwLock},
};

use kernel::{EntityGid, EntityKey, LookupBy};

pub trait EntityStorage {
    fn load(&self, lookup: &LookupBy) -> Result<Option<PersistedEntity>>;
    fn save(&self, entity: &PersistedEntity) -> Result<()>;
    fn delete(&self, entity: &PersistedEntity) -> Result<()>;
    fn begin(&self) -> Result<()>;
    fn rollback(&self, benign: bool) -> Result<()>;
    fn commit(&self) -> Result<()>;
    fn query_all(&self) -> Result<Vec<PersistedEntity>>;
}

pub trait EntityStorageFactory: Send + Sync {
    fn create_storage(&self) -> Result<Rc<dyn EntityStorage>>;
}

#[derive(Clone, Debug)]
pub struct PersistedEntity {
    pub key: String,
    pub gid: u64,
    pub version: u64,
    pub serialized: String,
}

impl PersistedEntity {
    pub fn to_json_value(&self) -> Result<serde_json::Value> {
        Ok(serde_json::from_str(&self.serialized)?)
    }
}

#[derive(Default)]
pub struct InMemoryEntityStorageFactory {
    entities: Arc<RwLock<HashMap<EntityKey, PersistedEntity>>>,
}

impl EntityStorageFactory for InMemoryEntityStorageFactory {
    fn create_storage(&self) -> Result<Rc<dyn EntityStorage>> {
        Ok(Rc::new(InMemoryEntityStorage {
            entities: self.entities.clone(),
            pending: Default::default(),
        }))
    }
}

enum Pending {
    Save(PersistedEntity),
    Delete(PersistedEntity),
}

pub struct InMemoryEntityStorage {
    entities: Arc<RwLock<HashMap<EntityKey, PersistedEntity>>>,
    pending: RwLock<Vec<Pending>>,
}

impl EntityStorage for InMemoryEntityStorage {
    fn load(&self, lookup: &LookupBy) -> Result<Option<PersistedEntity>> {
        let entities = self.entities.read().expect("Lock error");
        let entity = entities
            .iter()
            .filter(|(_, e)| match lookup {
                LookupBy::Key(key) => e.key == key.key_to_string(),
                LookupBy::Gid(gid) => EntityGid::new(e.gid) == **gid,
            })
            .map(|(_, e)| e)
            .next();

        Ok(entity.cloned())
    }

    fn save(&self, entity: &PersistedEntity) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        pending.push(Pending::Save(entity.clone()));

        Ok(())
    }

    fn delete(&self, entity: &PersistedEntity) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        pending.push(Pending::Delete(entity.clone()));

        Ok(())
    }

    fn begin(&self) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        pending.clear();

        Ok(())
    }

    fn rollback(&self, _benign: bool) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        pending.clear();

        Ok(())
    }

    fn commit(&self) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        let mut entities = self.entities.write().expect("Lock error");

        for pending in pending.iter() {
            match pending {
                Pending::Save(e) => entities.insert(EntityKey::new(&e.key), e.clone()),
                Pending::Delete(e) => entities.remove(&EntityKey::new(&e.key)),
            };
        }

        pending.clear();

        Ok(())
    }

    fn query_all(&self) -> Result<Vec<PersistedEntity>> {
        let entities = self.entities.read().expect("Lock error");

        Ok(entities.values().map(|e| e.clone()).collect())
    }
}
