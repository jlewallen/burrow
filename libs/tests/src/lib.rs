use anyhow::Result;
use engine::storage::{EntityStorageFactory, InMemoryEntityStorageFactory};
use plugins_dynlib::DynamicPluginFactory;
use plugins_rune::RunePluginFactory;
use plugins_wasm::WasmPluginFactory;
use replies::Reply;
use std::env::temp_dir;
use std::rc::Rc;
use std::sync::Arc;

use engine::{sequences::DeterministicKeys, DevNullNotifier, Domain, Session, SessionOpener};
use kernel::RegisteredPlugins;
use plugins_core::building::BuildingPluginFactory;
use plugins_core::carrying::CarryingPluginFactory;
use plugins_core::looking::LookingPluginFactory;
use plugins_core::moving::MovingPluginFactory;
use plugins_core::DefaultFinder;
use plugins_core::{BuildSurroundings, QuickThing};
// use sqlite::Factory;

pub const USERNAME: &str = "burrow";

pub trait WorldFixture {
    fn prepare(&self, session: &Rc<Session>) -> Result<()>;
}

#[derive(Default)]
pub struct Noop {}

impl WorldFixture for Noop {
    fn prepare(&self, _session: &Rc<Session>) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
pub struct HoldingKeyInVessel {}

impl WorldFixture for HoldingKeyInVessel {
    fn prepare(&self, session: &Rc<Session>) -> Result<()> {
        let mut build = BuildSurroundings::new_in_session(session.clone())?;

        let place = build.make(QuickThing::Place("Place"))?;
        let key = build.entity()?.named("Key")?.into_entry()?;
        let vessel = build
            .entity()?
            .named("Vessel")?
            .holding(&vec![key.clone()])?
            .into_entry()?;
        let (_, _surroundings) = build
            .route("East", QuickThing::Actual(place.clone()))
            .hands(vec![QuickThing::Actual(vessel.clone())])
            .build()?;

        session.flush()?;

        Ok(())
    }
}

fn evaluate_fixture<W, S>(
    domain: &S,
    username: &str,
    text: &'static [&'static str],
) -> Result<Option<Box<dyn Reply>>>
where
    W: WorldFixture + Default,
    S: SessionOpener,
{
    let session = domain.open_session()?;

    let fixture = W::default();

    fixture.prepare(&session)?;

    for text in text {
        if let Some(_reply) = session.evaluate_and_perform(username, text)? {
            // Do nothing, for now.
        }
    }

    session.close(&DevNullNotifier::default())?;

    Ok(None)
}

fn make_domain() -> Result<Domain> {
    test_domain_with(InMemoryEntityStorageFactory::default())
}

pub fn evaluate_text_in_new_domain<W>(
    username: &str,
    times: usize,
    text: &'static [&'static str],
) -> Result<()>
where
    W: WorldFixture + Default,
{
    let domain = make_domain()?;

    assert!(times > 0);

    evaluate_fixture::<W, _>(&domain, username, text)?;

    for _ in [0..times - 1] {
        evaluate_fixture::<Noop, _>(&domain, username, text)?;
    }

    Ok(())
}

pub fn test_domain_with<S>(storage: S) -> Result<Domain>
where
    S: EntityStorageFactory + 'static,
{
    let storage_factory = Arc::new(storage);
    let mut registered_plugins = RegisteredPlugins::default();
    registered_plugins.register(MovingPluginFactory::default());
    registered_plugins.register(LookingPluginFactory::default());
    registered_plugins.register(CarryingPluginFactory::default());
    registered_plugins.register(BuildingPluginFactory::default());
    registered_plugins.register(DynamicPluginFactory::default());
    registered_plugins.register(RunePluginFactory::default());
    registered_plugins.register(WasmPluginFactory::new(&temp_dir())?);
    let finder = Arc::new(DefaultFinder::default());
    let keys = Arc::new(DeterministicKeys::new());
    let identities = Arc::new(DeterministicKeys::new());
    Ok(Domain::new(
        storage_factory,
        Arc::new(registered_plugins),
        finder,
        keys,
        identities,
    ))
}

#[cfg(test)]
mod tests;
