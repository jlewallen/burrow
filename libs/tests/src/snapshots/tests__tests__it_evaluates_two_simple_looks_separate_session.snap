---
source: libs/tests/src/lib.rs
expression: domain.snapshot().await?
---
[
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-0",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 1
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Place"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {},
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-1",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 2
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Key"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "location": {
        "container": {
          "gid": 3,
          "key": "E-2",
          "klass": "",
          "name": "Vessel",
          "py/object": "model.entity.EntityRef",
          "py/ref": "model.entity.Entity"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-2",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 3
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Vessel"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "containing": {
        "capacity": null,
        "holding": [
          {
            "gid": 2,
            "key": "E-1",
            "klass": "",
            "name": "Key",
            "py/object": "model.entity.EntityRef",
            "py/ref": "model.entity.Entity"
          }
        ],
        "produces": {}
      },
      "location": {
        "container": {
          "gid": 4,
          "key": "E-3",
          "klass": "",
          "name": "Living",
          "py/object": "model.entity.EntityRef",
          "py/ref": "model.entity.Entity"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-3",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 4
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Living"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "containing": {
        "capacity": null,
        "holding": [
          {
            "gid": 3,
            "key": "E-2",
            "klass": "",
            "name": "Vessel",
            "py/object": "model.entity.EntityRef",
            "py/ref": "model.entity.Entity"
          }
        ],
        "produces": {}
      },
      "occupying": {
        "area": {
          "gid": 6,
          "key": "E-5",
          "klass": "",
          "name": "Welcome Area",
          "py/object": "model.entity.EntityRef",
          "py/ref": "model.entity.Entity"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "world",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 5
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "World"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "identifiers": {
        "acls": {
          "py/object": "",
          "rules": []
        },
        "gid": 7
      },
      "usernames": {
        "users": {
          "burrow": "E-3"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-5",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 6
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Welcome Area"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "containing": {
        "capacity": null,
        "holding": [
          {
            "gid": 7,
            "key": "E-6",
            "klass": "",
            "name": "East",
            "py/object": "model.entity.EntityRef",
            "py/ref": "model.entity.Entity"
          }
        ],
        "produces": {}
      },
      "occupyable": {
        "acls": {
          "py/object": "",
          "rules": []
        },
        "occupancy": 0,
        "occupied": [
          {
            "gid": 4,
            "key": "E-3",
            "klass": "",
            "name": "Living",
            "py/object": "model.entity.EntityRef",
            "py/ref": "model.entity.Entity"
          }
        ]
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-6",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 7
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "East"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "exit": {
        "area": {
          "gid": 1,
          "key": "E-0",
          "klass": "",
          "name": "Place",
          "py/object": "model.entity.EntityRef",
          "py/ref": "model.entity.Entity"
        }
      },
      "location": {
        "container": {
          "gid": 6,
          "key": "E-5",
          "klass": "",
          "name": "Welcome Area",
          "py/object": "model.entity.EntityRef",
          "py/ref": "model.entity.Entity"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  }
]
